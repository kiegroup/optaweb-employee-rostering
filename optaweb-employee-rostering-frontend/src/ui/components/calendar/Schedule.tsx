// @ts-nocheck
// TODO: Re-enable typescript validation on this file when
// @types/react-big-calendar is updated to have prop dayLayoutAlgorithm
import * as React from 'react';
import { Calendar, momentLocalizer, EventProps } from 'react-big-calendar';
import moment from 'moment';
import withDragAndDrop from 'react-big-calendar/lib/addons/dragAndDrop';
import { doNothing } from 'types';
import EventWrapper from './EventWrapper';

import 'react-big-calendar/lib/addons/dragAndDrop/styles.scss';
import './ReactBigCalendarOverrides.css';

const DragAndDropCalendar = withDragAndDrop(Calendar);

export interface StyleContainer {
  style?: React.CSSProperties;
  className?: string;
}

export type StyleSupplier<T> = (params: T) => StyleContainer;

export interface Props<T extends object> {
  startDate: Date;
  endDate: Date;
  events: T[];
  showAllDayCell?: boolean;
  dateFormat?: (date: Date) => string;
  startAccessor: (event: T) => Date;
  endAccessor: (event: T) => Date;
  titleAccessor: (event: T) => string;
  onAddEvent: (start: Date, end: Date) => void;
  onUpdateEvent: (event: T, start: Date, end: Date) => void;
  eventStyle: StyleSupplier<T>;
  wrapperStyle: StyleSupplier<T>;
  dayStyle: StyleSupplier<Date>;
  popoverHeader: (event: T) => React.ReactNode;
  popoverBody: (event: T) => React.ReactNode;
  eventComponent: (props: React.PropsWithChildren<EventProps<T>>) => React.ReactNode;
}

export function isDay(start: Date, end: Date) {
  return start.getHours() === 0 && start.getMinutes() === 0
    && end.getHours() === 0 && end.getMinutes() === 0;
}

const localizer = momentLocalizer(moment);
export default function Schedule<T extends object>(props: Props<T>): React.ReactElement<Props<T>> {
  const length = Math.ceil(moment(props.endDate).diff(moment(props.startDate), 'days')) + 1;
  const ref = React.useRef<HTMLElement>();
  return (
    <div
      style={{
        height: 'calc(100% - 20px)',
      }}
      ref={ref}
    >
      <DragAndDropCalendar
        className={(props.showAllDayCell) ? undefined : 'rbc-no-allday-cell'}
        dayLayoutAlgorithm="no-overlap"
        date={props.startDate}
        length={length}
        localizer={localizer}
        events={props.events}
        titleAccessor={props.titleAccessor}
        allDayAccessor={event => (props.showAllDayCell
          ? isDay(props.startAccessor(event), props.endAccessor(event)) : false)}
        startAccessor={props.startAccessor}
        endAccessor={props.endAccessor}
        toolbar={false}
        view="week"
        views={['week']}
        formats={props.dateFormat ? {
          dayFormat: props.dateFormat,
        } : undefined}
        onSelectSlot={(slotInfo: { start: string|Date; end: string|Date;
          action: 'select'|'click'|'doubleClick'; }) => {
          if (slotInfo.action === 'select' || slotInfo.action === 'click') {
            if (isDay(moment(slotInfo.start).toDate(), moment(slotInfo.end).toDate())) {
              props.onAddEvent(moment(slotInfo.start).toDate(), moment(slotInfo.end).add(1, 'day').toDate());
            } else {
              props.onAddEvent(moment(slotInfo.start).toDate(), moment(slotInfo.end).toDate());
            }
          }
        }
        }
        onEventDrop={(dropLocation: { event: T; start: string|Date; end: string|Date }) => {
          if (isDay(moment(dropLocation.start).toDate(), moment(dropLocation.end).toDate())) {
            props.onUpdateEvent(dropLocation.event, moment(dropLocation.start).toDate(),
              moment(dropLocation.end).toDate());
          } else if (moment(dropLocation.start).dayOfYear() !== moment(dropLocation.end).dayOfYear()) {
            props.onUpdateEvent(dropLocation.event, moment(dropLocation.start).toDate(), moment(dropLocation.start)
              .add(
                moment(props.endAccessor(dropLocation.event))
                  .diff(moment(props.startAccessor(dropLocation.event)),
                    'minutes'),
                'minutes',
              )
              .toDate());
          } else {
            props.onUpdateEvent(dropLocation.event, moment(dropLocation.start).toDate(), moment(dropLocation.end)
              .toDate());
          }
        }}
        onEventResize={(resizeInfo: { event: T; start: string|Date; end: string|Date }) => {
          const origEventStart = moment(props.startAccessor(resizeInfo.event));
          const origEventEnd = moment(props.startAccessor(resizeInfo.event));
          if (isDay(moment(resizeInfo.start).toDate(), moment(resizeInfo.end).toDate())) {
            props.onUpdateEvent(resizeInfo.event, moment(resizeInfo.start).toDate(), moment(resizeInfo.end)
              .toDate());
          } else if (origEventStart.dayOfYear() !== moment(resizeInfo.start).dayOfYear()) {
            props.onUpdateEvent(resizeInfo.event, origEventStart.toDate(), moment(resizeInfo.end).toDate());
          } else if (origEventEnd.dayOfYear() !== moment(resizeInfo.end).dayOfYear()) {
            props.onUpdateEvent(resizeInfo.event, moment(resizeInfo.start).toDate(), origEventEnd.toDate());
          } else {
            props.onUpdateEvent(resizeInfo.event, moment(resizeInfo.start).toDate(), moment(resizeInfo.end).toDate());
          }
        }}
        onView={doNothing}
        onNavigate={doNothing}
        timeslots={4}
        eventPropGetter={props.eventStyle}
        dayPropGetter={props.dayStyle}
        selectable
        resizable
        showMultiDayTimes
        components={{
          eventWrapper: (wrapperProps) => {
            const style = props.wrapperStyle(wrapperProps.event);
            return EventWrapper({
              ...wrapperProps,
              style: {
                ...wrapperProps.style,
                ...((style.style) ? style.style : {}),
              },
              boundary: ref,
              className: style.className ? style.className : '',
              popoverHeader: props.popoverHeader(wrapperProps.event),
              popoverBody: props.popoverBody(wrapperProps.event),
            });
          },
          event: eventProps => props.eventComponent(eventProps) as React.ReactElement,
        }}
      />
    </div>
  );
}
