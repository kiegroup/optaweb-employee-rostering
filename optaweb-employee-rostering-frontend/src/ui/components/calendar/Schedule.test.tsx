import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { CalendarProps, EventProps, EventWrapperProps } from 'react-big-calendar';
import { withDragAndDropProps } from 'react-big-calendar/lib/addons/dragAndDrop';
import Schedule, { isDay, Props } from './Schedule';

function calendarProps(wrapper: ShallowWrapper): withDragAndDropProps<{ start: Date; end: Date; title: string }> &
CalendarProps<{ start: Date; end: Date; title: string }> {
  return (wrapper.find('DragAndDropCalendar').props() as
    withDragAndDropProps<{ start: Date; end: Date; title: string }> &
    CalendarProps<{ start: Date; end: Date; title: string }>);
}

function reportError(prop: string): never {
  throw new Error(`${prop}  should be defined.`);
}

describe('Schedule', () => {
  it('isDay should return true if both start and end are on midnight', () => {
    const startDate = new Date('2018-01-01T00:00');
    const endDate = new Date('2018-01-02T00:00');
    expect(isDay(startDate, endDate)).toEqual(true);
  });

  it('isDay should return false if either start or end are not midnight', () => {
    const startDate1 = new Date('2018-01-01T12:00');
    const endDate1 = new Date('2018-01-02T00:00');
    expect(isDay(startDate1, endDate1)).toEqual(false);

    const startDate2 = new Date('2018-01-01T00:00');
    const endDate2 = new Date('2018-01-02T12:00');
    expect(isDay(startDate2, endDate2)).toEqual(false);
  });

  it('should render correctly without an all-day cell', () => {
    const schedule = shallow(<Schedule {...props} />);
    const allDayAccessor: Function = calendarProps(schedule).allDayAccessor as Function;
    expect(allDayAccessor(props.events[0])).toEqual(false);
    expect(allDayAccessor(props.events[1])).toEqual(false);
    expect(allDayAccessor(props.events[2])).toEqual(false);
  });

  it('should render correctly with an all-day cell', () => {
    const schedule = shallow(<Schedule {...props} showAllDayCell />);
    const allDayAccessor: Function = calendarProps(schedule).allDayAccessor as Function;
    expect(allDayAccessor(props.events[0])).toEqual(false);
    expect(allDayAccessor(props.events[1])).toEqual(true);
    expect(allDayAccessor(props.events[2])).toEqual(false);
  });

  it('length should be correct', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { length } = calendarProps(schedule);
    expect(length).toEqual(7);
  });

  it('title accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { titleAccessor } = calendarProps(schedule);
    expect(titleAccessor).toBe(props.titleAccessor);
  });

  it('start accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { startAccessor } = calendarProps(schedule);
    expect(startAccessor).toBe(props.startAccessor);
  });

  it('end accessor should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { endAccessor } = calendarProps(schedule);
    expect(endAccessor).toBe(props.endAccessor);
  });

  it('day format should be date format if provided', () => {
    const dateFormat = jest.fn();
    const schedule = shallow(<Schedule {...props} dateFormat={dateFormat} />);
    const { dayFormat } = calendarProps(schedule).formats || reportError('formats');
    expect(dayFormat).toBe(dateFormat);
  });

  it('formats should be undefined if date format not provided', () => {
    const schedule = shallow(<Schedule {...props} />);
    expect(calendarProps(schedule).formats).toBeUndefined();
  });

  it('should not create event on double click', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError('onSelectSlot');
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: 'doubleClick',
    });
    expect(props.onAddEvent).not.toBeCalled();
  });

  it('should create event on click', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError('onSelectSlot');
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: 'click',
    });
    expect(props.onAddEvent).toBeCalled();
    expect(props.onAddEvent).toBeCalledWith(props.events[0].start, props.events[0].end);
  });

  it('should create event on select', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError('onSelectSlot');
    onSelectSlot({
      start: props.events[0].start,
      end: props.events[0].end,
      slots: [],
      action: 'select',
    });
    expect(props.onAddEvent).toBeCalled();
    expect(props.onAddEvent).toBeCalledWith(props.events[0].start, props.events[0].end);
  });

  it('should move event on drag', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onEventDrop = calendarProps(schedule).onEventDrop || reportError('onEventDrop');
    onEventDrop({
      event: props.events[0],
      start: props.events[0].start,
      end: props.events[0].end,
      allDay: false,
    });
    expect(props.onUpdateEvent).toBeCalled();
    expect(props.onUpdateEvent).toBeCalledWith(props.events[0], props.events[0].start, props.events[0].end);
  });

  it('should resize event on resize', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onEventResize = calendarProps(schedule).onEventResize || reportError('onEventResize');
    onEventResize({
      event: props.events[0],
      start: props.events[0].start,
      end: props.events[0].end,
      allDay: false,
    });
    expect(props.onUpdateEvent).toBeCalled();
    expect(props.onUpdateEvent).toBeCalledWith(props.events[0], props.events[0].start, props.events[0].end);
  });

  it('should create event of proper length if it span a day', () => {
    const schedule = shallow(<Schedule {...props} />);
    const onSelectSlot = calendarProps(schedule).onSelectSlot || reportError('onSelectSlot');
    onSelectSlot({
      start: props.events[1].start,
      end: props.events[1].start, // Not an error, selecting a single day yields start and end the same
      slots: [],
      action: 'click',
    });
    expect(props.onAddEvent).toBeCalled();
    expect(props.onAddEvent).toBeCalledWith(props.events[1].start, props.events[1].end);
  });

  it('eventPropGetter should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { eventPropGetter } = calendarProps(schedule);
    expect(eventPropGetter).toBe(props.eventStyle);
  });

  it('dayPropGetter should be prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const { dayPropGetter } = calendarProps(schedule);
    expect(dayPropGetter).toBe(props.dayStyle);
  });

  it('event component should defer to prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const components = calendarProps(schedule).components || reportError('components');
    const event = components.event as
      React.FC<React.PropsWithChildren<EventProps<{ start: Date; end: Date; title: string }>>>
      || reportError('components.event');
    const out = event(
      {
        event: props.events[0],
        ...props.events[0],
      },
    );
    expect(props.eventComponent).toBeCalled();
    expect(props.eventComponent).toBeCalledWith({
      event: props.events[0],
      ...props.events[0],
    });
    expect(out).toEqual(props.eventComponent({
      event: props.events[0],
      ...props.events[0],
    }));
  });

  it('wrapper component should defer to prop', () => {
    const schedule = shallow(<Schedule {...props} />);
    const components = calendarProps(schedule).components || reportError('components');
    const wrapper = components.eventWrapper as
      React.FC<React.PropsWithChildren<EventWrapperProps<{ start: Date; end: Date; title: string }>>>
      || reportError('components.event');
    const wrapperProps: React.PropsWithChildren<EventWrapperProps<{ start: Date; end: Date; title: string }>> = {
      event: props.events[0],
      className: 'wrapperClass',
      isRtl: false,
      getters: {},
      onClick: jest.fn(),
      onDoubleClick: jest.fn(),
      accessors: {},
      selected: false,
      label: 'Label',
      continuesEarlier: false,
      continuesLater: true,
      style: {
        top: '33%',
        height: '67%',
        xOffset: 0,
      },
    };
    const out = wrapper(wrapperProps);
    expect(props.wrapperStyle).toBeCalled();
    expect(props.wrapperStyle).toBeCalledWith(props.events[0]);
    expect(out).toMatchSnapshot();
  });
});

const props: Props<{ start: Date; end: Date; title: string }> = {
  startDate: new Date('2018-01-01T00:00'),
  endDate: new Date('2018-01-07T00:00'),
  events: [
    {
      title: 'Event 1',
      start: new Date('2018-01-02T08:00'),
      end: new Date('2018-01-02T17:00'),
    },
    {
      title: 'All Day Event ',
      start: new Date('2018-01-03T00:00'),
      end: new Date('2018-01-04T00:00'),
    },
    {
      title: 'Multi-day Event',
      start: new Date('2018-01-0520:00'),
      end: new Date('2018-01-06T06:00'),
    },
  ],
  showAllDayCell: false,
  startAccessor: jest.fn(e => e.start),
  endAccessor: jest.fn(e => e.end),
  titleAccessor: jest.fn(e => e.title),
  onAddEvent: jest.fn(),
  onUpdateEvent: jest.fn(),
  eventStyle: jest.fn(() => ({ style: { color: 'red' }, className: 'class' })),
  wrapperStyle: jest.fn(() => ({ style: { color: 'red' }, className: 'class', props: { prop: true } })),
  dayStyle: jest.fn(() => ({ style: { color: 'red' }, className: 'class' })),
  popoverHeader: jest.fn(() => 'Title'),
  popoverBody: jest.fn(() => 'Body'),
  eventComponent: jest.fn(e => e.title),
};
