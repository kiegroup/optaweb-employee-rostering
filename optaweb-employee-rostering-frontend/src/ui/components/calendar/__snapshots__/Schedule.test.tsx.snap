// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule wrapper component should defer to prop 1`] = `
<div
  className="class continues-next-day"
  style={
    Object {
      "backgroundColor": "transparent",
      "border": "none",
      "color": "red",
      "height": "67%",
      "top": "33%",
      "xOffset": 0,
    }
  }
>
  <Popover
    bodyContent="Body"
    headerContent="Title"
    zIndex={1000001}
  />
</div>
`;
